File: server/php/w2lib.php
<?php
/**************************************************
*	This class helps streamline w2ui requests
* 
*	Global Variables (it is assumed that there are several global variables)
*	- $db - connection to the database
* 	- $dbType - type of the db, can be mysql or postgres
*/

$w2grid = new w2grid_class();

class w2grid_class {
	// constructor/destructor
	public function __construct() {}
	public function __destruct() {}

	public function getRecords($sql, $cql, $request) {
		global $db, $dbType;

		// prepare search
		$str = "";
		if (isset($request['search']) && is_array($request['search'])) {
			foreach ($request['search'] as $s => $search) {
				if ($str != "") $str .= " ".$request['search-logic']." ";
				$operator = "=";
				$field 	  = $search['field'];
				$value    = "'".$search['value']."'";
				switch (strtolower($search['operator'])) {

					case 'begins':
						$operator = ($dbType == "postgres" ? "ILIKE" : "LIKE");
						$value 	  = "'".$search['value']."%'";
						break;

					case 'ends':
						$operator = ($dbType == "postgres" ? "ILIKE" : "LIKE");
						$value 	  = "'%".$search['value']."'";
						break;

					case 'contains':
						$operator = ($dbType == "postgres" ? "ILIKE" : "LIKE");
						$value 	  = "'%".$search['value']."%'";
						break;

					case 'is':
						$operator = "=";
						if (!is_int($search['value']) && !is_float($search['value'])) {
							$field = "LOWER($field)";
							$value = "LOWER('".$search['value']."')";
						} else {
							$value = "'".$search['value']."'";
						}
						break;

					case 'between':
						$operator = "BETWEEN";
						$value 	  = "'".$search['value'][0]."' AND '".$search['value'][1]."'";
						break;

					case 'in':
						$operator = "IN";
						$value 	  = "[".$search['value']."]";
						break;
				}
				$str .= $field." ".$operator." ".$value;
			}
		}
		if ($str == "") $str = " 1=1 ";

		// prepare sort
		$str2 = "";
		if (isset($request['sort']) && is_array($request['sort'])) {
			foreach ($request['sort'] as $s => $sort) {
				if ($str2 != "") $str2 .= ", ";
				$str2 .= $sort['field']." ".$sort['direction'];
			}
		}
		if ($str2 == "") $str2 = "1=1";

		// build sql
		$sql = str_ireplace("~search~", $str, $sql);
		$sql = str_ireplace("~order~", "~sort~", $sql);
		$sql = str_ireplace("~sort~", $str2, $sql);

		// build cql (for counging)
		if ($cql == null || $cql == "") {
			$cql = "SELECT count(1) FROM ($sql) as grid_list_1";
		}
		if (!isset($request['limit']))  $request['limit']  = 50;
		if (!isset($request['offset'])) $request['offset'] = 0;

		$sql .= " LIMIT ".$request['limit']." OFFSET ".$request['offset'];

		$data = Array();		

		// count records
		$rs = $db->execute($cql);
		$data['status'] = 'success';
		$data['total']  = $rs->fields[0];

		// execute sql
		$rs = $db->execute($sql);
		//print($sql);

		// check for error
		if ($db->res_errMsg != '') {
			$data = Array();
			$data['status'] = 'error';
			$data['message'] = $db->res_errMsg;
			return $data;
		}
		$data['records'] = array();

		$len = 0;
		while($rs && !$rs->EOF) {
			$data['records'][$len] = Array();
			$data['records'][$len]['recid'] = $rs->fields[0];
			foreach ($rs->fields as $k => $v) {
				if (intval($k) > 0 || $k == "0") continue;
				$data['records'][$len][$k] = $v;
			}
			$len++;
			$rs->moveNext();
		}
		return $data;
	}

	public function deleteRecords($table, $keyField, $data) {
		global $db;
		$res = Array();

		$recs = "";
		foreach ($data['selected'] as $k => $v) {
			if ($recs != "") $recs .= ", ";
			$recs .= "'".addslashes($v)."'";
		}
		$sql = "DELETE FROM $table WHERE $keyField IN ($recs)";
		$rs = $db->execute($sql);
		// check for error
		if ($db->res_errMsg != '') {
			$res['status'] = 'error';
			$res['message'] = $db->res_errMsg;
			return $res;
		}
		$res['status']  = 'success';
		$res['message'] = '';
		return $res;
	}

	public function getRecord($sql) {
		global $db;
		$data = Array();

		// execute sql
		$rs = $db->execute($sql);
		// check for error
		if ($db->res_errMsg != '') {
			$data = Array();
			$data['status'] = 'error';
			$data['message'] = $db->res_errMsg;
			return $data;
		}

		$data['status'] = 'success';
		$data['record']	= Array();
		while ($rs && !$rs->EOF) {
			foreach ($rs->fields as $k => $v) {
				if (intval($k) > 0 || $k == "0") continue;
				$data['record'][$k] = $v;
			}
			break;
		}
		return $data;
	}

	public function saveRecord($table, $keyField, $data) {
		global $db;

		if ($data['recid'] == '' || $data['recid'] == '0') {
			$fields = "";
			$values = "";
			foreach ($data['record'] as $k => $v) {
				if ($k == $keyField) continue; // key field should not be here
				if ($fields != '') $fields .= ", ";
				if ($values != '') $values .= ", ";
				$fields .= addslashes($k);
				if (substr($v, 0, 2) == "__") {
					$values .= addslashes(substr($v, 2));
				} else {
					$values .= ($v == "" ? "null" : "'".addslashes($v)."'");
				}
			}
			$sql = "INSERT INTO $table($fields) VALUES($values)";
		} else {
			$values = "";
			foreach ($data['record'] as $k => $v) {
				if ($k == $keyField) continue; // key field should not be here
				if ($values != '') $values .= ", ";
				if (substr($v, 0, 2) == "__") {
					$values .= addslashes($k)." = ".addslashes(substr($v, 2));
				} else {
					$values .= addslashes($k)." = ".($v == "" ? "null" : "'".addslashes($v)."'");
				}
			}
			$sql = "UPDATE $table SET $values WHERE $keyField = ".addslashes($data['recid']);
		}
		// execute sql
		$rs = $db->execute($sql);
		// check for error
		if ($db->res_errMsg != '') {
			$res = Array();
			$res['status'] = 'error';
			$res['message'] = $db->res_errMsg;
			return $res;
		}

		$res = Array();
		$res['status']  = 'success';
		$res['message'] = '';
		return $res;
	}

	public function newRecord($table, $data) {
		global $db;

		$res    = Array();
		$fields = '';
		$values = '';

		foreach ($data as $k => $v) {
			if ($fields != '') $fields .= ",";
			if ($values != '') $values .= ",";
			$fields .= $k;
			if (substr($v, 0, 2) == "__") {
				$values .= addslashes(substr($v, 2));
			} else {
				$values .= ($v == "" ? "null" : "'".addslashes($v)."'");
			}
		}

		$sql = "INSERT INTO $table($fields) VALUES ($values)";
		$db->execute($sql);
		if ($db->res_errMsg != '') {
			$res['status']  = 'error';
			$res['message'] = $db->res_errMsg;
		} else {
			$res['status']  = 'success';
		}
		return $res;	
	}	

	public function getItems($sql) {
		global $db;
		$data = Array();

		// execute sql
		$rs = $db->execute($sql);
		// check for error
		if ($db->res_errMsg != '') {
			$data = Array();
			$data['status'] = 'error';
			$data['message'] = $db->res_errMsg;
			return $data;
		}

		$len = 0;
		$data['status']  = 'success';
		$data['total'] 	 = $db->res_rowCount;
		$data['items'] = Array();
		while ($rs && !$rs->EOF) {
			$data['items'][$len] = Array();
			$data['items'][$len]['id']   = $rs->fields[0];
			$data['items'][$len]['text'] = $rs->fields[1];
			foreach ($rs->fields as $k => $v) {
				if (intval($k) > 0 || $k == "0") continue;
				$data['items'][$len][$k] = $v;
			}
			$len++;
			if ($len >= $_REQUEST['max']) break;
			$rs->moveNext();
		}
		return $data;
	}

	public function outputJSON($data) {
		header("Content-Type: application/json;charset=utf-8");
		echo json_encode($data);
	}
}

File: server/python/django_w2ui/django_w2ui/w2lib.py
class w2Grid:
  def __init__(self,conn):
    self.conn = conn

  def getRecords(self, sql, request, cql=None):
    sql_components = { 'where': [], 'params': [], 'sort': [] }
    if request.get('search',[]):
      for search in request['search']:
        operator = "="
        field    = search['field']  # TODO: protect from sql injection!!!
        value    = [ search['value'] ]
        op       = search['operator'].lower()
        if op == "begins":
          operator = "LIKE ?||'%%'"
        elif op == "ends":
          operator = "LIKE '%%'||?"
        elif op == "contains":
          operator = "LIKE '%%'||?||'%%'"
        elif op == "is":
          operator = "= LOWER(?)"
        elif op == "between":
          value    = value[0]
          operator = "BETWEEN ? AND ?"
        elif op == "in":
          value    = value[0]
          operator = "IN (%s)" % ','.join(['?'] * len(value))
        sql_components['where'].append("%s %s" % (field,operator))
        for v in value:
          sql_components['params'].append(v)

    if request.get('sort',[]):
      for sort in request['sort']:
         field = sort['field']      # TODO: protect from sql injection!!!
         dir_  = sort['direction']  # TODO: protect from sql injection!!!
         sql_components['sort'].append(field+' '+dir_)

    connector = ' %s ' % request.get('searchLogic','AND')  # TODO: protect from sql injection!!!
    where = connector.join(sql_components['where'])
    if not where:
      where = '1=1'
    sort = ",".join(sql_components['sort'])
    if not sort:
      sort = '1'
    sql = sql.replace("~search~",where)
    sql = sql.replace("~order~","~sort~")
    sql = sql.replace("~sort~",sort)

    if not cql:
      cql = "SELECT count(1) FROM (%s) as grid_list_1" % sql

    limit  = 50
    offset = 0
    try:
      limit = abs(int(request['limit']))
    except:
      pass
    try:
      offset = abs(int(request['offset']))
    except:
      pass
    sql += " LIMIT %s OFFSET %s" % (limit,offset)

    data = {}
    try:
      cursor = self.conn.cursor()
      # count records
      cursor.execute(cql,sql_components['params'])
      data['status'] = 'success'
      data['total'] = cursor.fetchone()[0]

      # execute sql
      data['records'] = []
      rows = cursor.execute(sql,sql_components['params'])
      columns = [ d[0] for d in cursor.description ]
      columns[0] = "recid"
      for row in rows:
        record = zip(columns,list(row))
        data['records'].append( dict(record) )
    except Exception, e:
      data['status'] = 'error'
      data['message'] = '%s\n%s' % (e,sql)
    return data

  def deleteRecords(self, table, keyField, request):
    recs = request['selected'] 
    # TODO: protect table, keyField from sql injection!!!
    sql = "DELETE FROM %s WHERE %s IN (%s)" % (table, keyField,','.join(['?'] * len(recs)))
    data = {}
    try:
      cursor = self.conn.cursor()
      cursor.execute(sql,recs)
      self.conn.commit()
      data['status']  = 'success'
      data['message'] = ''
    except Exception, e:
      data['status'] = 'error'
      data['message'] = '%s\n%s' % (e,sql)
    return data

  def getRecord(self, sql, recid):
    data = {}
    try:
      cursor = self.conn.cursor()
      # execute sql
      cursor.execute(sql,[recid])
      data['status']  = 'success'
      data['message'] = ''
      columns = [ d[0] for d in cursor.description ]
      row = cursor.fetchone()
      record = zip(columns,list(row))[1:]
      data['record'] = dict(record)
    except Exception, e:
      data['status'] = 'error'
      data['message'] = '%s\n%s' % (e,sql)
    return data

  def saveRecord(self, table, keyField, request):
    # TODO: protect table, keyField, field names from sql injection!!!
    fields, values = [], []
    for k, v in request['record'].items():
      if k == keyField: continue # key field should not be here
      fields.append(k)
      if v.startswith('__'):
        v = v[2:]
      elif v == "":
        v = None
      values.append(v)

    if request.get('recid','0') == '0':
      sql = "INSERT INTO %s (%s) VALUES (%s)" % (table,','.join(fields),','.join(['?']*len(fields)))
    else:
      sql = "UPDATE %s SET %s WHERE %s = ?" % (table, ','.join([ '%s=?' % f for f in fields ]), keyField)
      values.append( request['recid'] )
        
    data = {}
    try:
      cursor = self.conn.cursor()
      cursor.execute(sql,values)
      self.conn.commit()
      data['status']  = 'success'
      data['message'] = ''
    except Exception, e:
      data['status'] = 'error'
      data['message'] = '%s\n%s' % (e,sql)
    return data

  def newRecord(self, table, data):
    return self.saveRecord(table, '', {'recid': 0, 'record': data})

  def getItems(self, sql):
    # TODO: what's this function for?
    return {}
